imitation/model_env.py

```diff
@@ -0,0 +1,346 @@
+                                                 shape=(self.obs_dim, ))
+        return self._observation_space
+
+    def seed(self, seed=None):
+        if seed is None:
+            seed = np.random.randint(0, 1 << 31)
+        self.rand_state = np.random.RandomState(seed=seed)
+        return [seed]
+
+    def reset(self):
+        self.cur_state = 0
+        self.actions_taken = 0
+        # as in step(), we copy so that it can't be mutated in-place (updates
+        # will be reflected in self.observation_matrix!)
+        return self.observation_matrix[self.cur_state].copy()
+
+    def step(self, action):
+        old_state = self.cur_state
+        out_dist = self.transition_matrix[old_state, action]
+        choice_states = np.arange(self.n_states)
+        next_state = int(
+            self.rand_state.choice(choice_states, p=out_dist, size=()))
+        self.cur_state = next_state
+        self.actions_taken += 1
+        done = self.actions_taken >= self.horizon
+        reward = self.reward_matrix[old_state]
+        assert np.isscalar(reward), reward
+        # copy so that it can't be mutated in-place
+        obs = self.observation_matrix[next_state].copy()
+        assert obs.ndim == 1, obs.shape
+        infos = {"old_state": old_state, "new_state": next_state}
+        return obs, reward, done, infos
+
+    @property
+    def n_states(self):
+        """Number of states in this MDP (int)."""
+        return self.transition_matrix.shape[0]
+
+    @property
+    def n_actions(self):
+        """Number of actions in this MDP (int)."""
+        return self.transition_matrix.shape[1]
+
+    @property
+    def obs_dim(self):
+        """Size of observation vectors for this MDP."""
+        return self.observation_matrix.shape[-1]
+
+    # ############################### #
+    # METHODS THAT MUST BE OVERRIDDEN #
+    # ############################### #
+
+    @property
+    @abc.abstractmethod
+    def transition_matrix(self):
+        """Yield a 3D transition matrix with dimensions corresponding to
+        current state, current action, and next state (in that order). In other
+        words, if `T` is our returned matrix, then `T[s,a,sprime]` is the
+        chance of transitioning into state `sprime` after taking action `a` in
+        state `s`."""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def observation_matrix(self):
+        """Yields 2D observation matrix with dimensions corresponding to
+        current state (first dim) and elements of observation (second dim)."""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def reward_matrix(self):
+        """Yields 2D reward matrix with dimensions corresponding to current
+        state and current action."""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def horizon(self):
+        """Number of actions that can be taken in an episode."""
+        pass
+
+
+def make_random_trans_mat(
```

Comment:
Do you mean an `examples` subdir at the root, or an `examples` package in `imitation`? I'd like to have them somewhere accessible so that we have simple, pre-tested environments to run MCE IRL on. At the moment one of those environments is unit tested, and both appear in the demo notebook for MCE IRL.

Reply:
Given that you're using them in the notebook, an `examples` package in `imitation` would make more sense. I think since `RandomMDP` is used in test cases probably OK to leave it in `model_env`.

What I want to avoid is there being a large amount of code that is potentially useful (so shouldn't just be deleted) but is not essential to the operation of the algorithms and is not tested / maintained.